### PR TITLE
PW-818: Fix for Magento 2.3 compatibility

### PR DESCRIPTION
**Pull request by Aleffio**
**Description**
Add isAjax to the POST parameters, this will allow checks from M2.3 to go through.
This will keep the compatibility with older Magento versions.

**Tested scenarios**
Notifications on M2.1 and M2.3
wrong credentials, ?version=true

**Fixed issue**:  #327

### CODE REVIEW

File: Controller/Process/Json.php [hunk 1]
```diff
@@ -69,6 +74,13 @@ public function __construct(
         $this->_resultFactory = $context->getResultFactory();
         $this->_adyenHelper = $adyenHelper;
         $this->_adyenLogger = $adyenLogger;
+        $this->postParams = json_decode(file_get_contents('php://input'), true);
+        
+        // Fix for Magento2.3 adding isAjax to the request params
+        if (interface_exists("\Magento\Framework\App\CsrfAwareActionInterface") && !isset($this->getRequest()->getParams()['version'])) {
+            $this->postParams['isAjax'] = true;
+            $this->_request->setParams($this->postParams);
```
Comment (lancergr) [inline Controller/Process/Json.php]: wouldn't $this->_request->setParam('isAjax', true) be sufficient?

File: Controller/Process/Json.php [hunk 2]
```diff
@@ -69,6 +74,13 @@ public function __construct(
         $this->_resultFactory = $context->getResultFactory();
         $this->_adyenHelper = $adyenHelper;
         $this->_adyenLogger = $adyenLogger;
+        $this->postParams = json_decode(file_get_contents('php://input'), true);
+        
+        // Fix for Magento2.3 adding isAjax to the request params
+        if (interface_exists("\Magento\Framework\App\CsrfAwareActionInterface") && !isset($this->getRequest()->getParams()['version'])) {
```
Comment (lancergr) [inline Controller/Process/Json.php]: better to change similar to magento implementation https://github.com/magento/magento2/blob/2.3.0/app/code/Magento/Backend/App/Request/BackendValidator.php#L99
```suggestion
if(interface_exists("\Magento\Framework\App\CsrfAwareActionInterface")) {
  $request = $this->getRequest();
  if ($request instanceof HttpRequest && $request->isPost()) {
        $request->setParam('isAjax', true);
   }
 }
```
and
```
use Magento\Framework\App\Request\Http as HttpRequest;
```

**Comment by pitbulk** [comment]
I tested and the following trick
```
        // CSRF Magento2.3 compatibility
        if (interface_exists("\Magento\Framework\App\CsrfAwareActionInterface")) {
            $request = $this->getRequest();
            if ($request instanceof HttpRequest && $request->isPost()) {
                $request->setParam('isAjax', true);
            }
        }
```
applied on the Action constructor will also work.
And that solution is compatible with previous versions of Magento.

**Comment by pitbulk** [comment]
Another solution, inject on the fly a valid Key
```
        // CSRF Magento2.3 compatibility
        if (interface_exists("\Magento\Framework\App\CsrfAwareActionInterface")) {
            $request = $this->getRequest();
            if ($request instanceof HttpRequest && $request->isPost() && empty($request->getParam('form_key'))) {
                $formKey = $this->_objectManager->get(\Magento\Framework\Data\Form\FormKey::class);
                $request->setParam('form_key', $formKey->getFormKey());
            }
        }
```

**Pull request merged**